pkg/controller/managedresources/secret_controller.go

```diff
@@ -0,0 +1,124 @@
+func (r *SecretReconciler) InjectStopChannel(stopCh <-chan struct{}) error {
+	r.ctx = utils.ContextFromStopChannel(stopCh)
+	return nil
+}
+
+// NewSecretReconciler creates a new secret reconciler.
+func NewSecretReconciler(log logr.Logger, class *ClassFilter) *SecretReconciler {
+	return &SecretReconciler{
+		log:   log,
+		class: class,
+	}
+}
+
+// Reconcile implements `reconcile.SecretReconciler`.
+func (r *SecretReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("secret", req)
+
+	secret := &corev1.Secret{}
+	if err := r.client.Get(r.ctx, req.NamespacedName, secret); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Stopping reconciliation of Secret, as it has been deleted")
+			return reconcile.Result{}, nil
+		}
+		log.Error(err, "Could not fetch Secret")
+		return reconcile.Result{}, err
+	}
+
+	resourceList := &resourcesv1alpha1.ManagedResourceList{}
+	if err := r.client.List(r.ctx, resourceList, client.InNamespace(secret.Namespace)); err != nil {
+		log.Error(err, "Could not fetch ManagedResources in namespace of ManagedResource")
+		return reconcile.Result{}, err
+	}
+
+	// check if there is at least one ManagedResource this controller is responsible for and which references this secret
+	secretIsReferenced := false
+	for _, resource := range resourceList.Items {
+		for _, ref := range resource.Spec.SecretRefs {
+			// check if we are responsible for this MR, class might have changed, then we need to remove our finalizer
+			if ref.Name == secret.Name && r.class.Responsible(&resource) {
+				secretIsReferenced = true
+				break
+			}
+		}
+	}
+
+	controllerFinalizer := r.class.FinalizerName()
+	secretFinalizers := sets.NewString(secret.Finalizers...)
+	addFinalizer, removeFinalizer := false, false
+	if secretIsReferenced && !secretFinalizers.Has(controllerFinalizer) {
+		addFinalizer = true
+		log.Info("adding finalizer to secret because it is referenced by a ManagedResource",
+			"finalizer", controllerFinalizer)
+	} else if !secretIsReferenced && secretFinalizers.Has(controllerFinalizer) {
+		removeFinalizer = true
+		log.Info("removing finalizer from secret because it is not referenced by a ManagedResource of this class",
+			"finalizer", controllerFinalizer)
+	}
+
+	if addFinalizer || removeFinalizer {
+		if err := utils.TryUpdate(r.ctx, retry.DefaultBackoff, r.client, secret, func() error {
+			secretFinalizers := sets.NewString(secret.Finalizers...)
+			if addFinalizer {
+				secretFinalizers.Insert(controllerFinalizer)
+			} else if removeFinalizer {
+				secretFinalizers.Delete(controllerFinalizer)
+			}
+			secret.Finalizers = secretFinalizers.UnsortedList()
+			return nil
+		}); client.IgnoreNotFound(err) != nil {
+			r.log.Error(err, "failed to update finalizers of secret")
+			// dont' run into exponential backoff for adding/removing finalizers
+			return reconcile.Result{RequeueAfter: 5 * time.Second}, err
```

Comment:
I think when err is non-nil, the RequeueAfter is ignored. Ref https://github.com/kubernetes-sigs/controller-runtime/blob/198197fe0fd7553cadd8afcd3338f8ee4b8a7e3f/pkg/internal/controller/controller.go#L256-L275

Reply:
You're right, good catch. Thanks for fixing!